File: freeciv-web/src/main/webapp/javascript/webgl/city.js
/**********************************************************************
    FreecivX.net - the web version of Freeciv. http://www.FreecivX.net/
    Copyright (C) 2009-2022  The FreecivX.net project

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU Affero General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Affero General Public License for more details.

    You should have received a copy of the GNU Affero General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.

***********************************************************************/


var city_worked_positions = {};
var city_labels_visible = true;

/****************************************************************************
Show labels with worked city tiles.
****************************************************************************/
function show_city_worked_tiles()
{
  if (active_city == null) return;

  if (city_tile(active_city) != null && city_label_positions[city_tile(active_city)['index']] != null) {
    city_label_positions[city_tile(active_city)['index']].visible = false;
  }
  let other_city_material = new THREE.MeshBasicMaterial( { color: 0xc33b3b, transparent: true, opacity: 0.4} );

  for (var tile_id in tiles) {
    var ptile = tiles[tile_id];
    if (active_city != null && ptile != null && ptile['worked'] != null
          && active_city['id'] == ptile['worked'] && active_city['output_food'] != null) {
      let ctile = city_tile(active_city);
      let d = map_distance_vector(ctile, ptile);
      let idx = get_city_dxy_to_index(d[0], d[1], active_city);
      let pos = map_to_scene_coords(ptile['x'], ptile['y']);
      let height = 5 + ptile['height'] * 100;
      if (ctile['index'] == ptile['index']) {
        height += 15;
      }

      let food_output = active_city['output_food'][idx];
      let shield_output = active_city['output_shield'][idx];
      let trade_output = active_city['output_trade'][idx];

      /* The ruleset may use large values scaled down to get greater
       * granularity. */
      food_output = Math.floor(food_output / game_info.granularity);
      shield_output = Math.floor(shield_output / game_info.granularity);
      trade_output = Math.floor(trade_output / game_info.granularity);

      if (city_worked_positions[ptile['index']] == null) {
        var mesh = create_city_worked_sprite(food_output, shield_output, trade_output);
        city_worked_positions[ptile['index']] = mesh;
        mesh.position.set(pos['x'] + 0, height + 10, pos['y'] - 4);
        if (scene != null) {
          scene.add(mesh);
        }
      }
    } else if (active_city != null && ptile != null && ptile['worked'] != null
                && active_city['id'] != ptile['worked'] && ptile['worked'] > 0) {
      // tile worked by other city
      let ctile = city_tile(active_city);
      let d = map_distance_vector(ctile, ptile);
      let idx = get_city_dxy_to_index(d[0], d[1], active_city);
      let pos = map_to_scene_coords(ptile['x'], ptile['y']);
      let height = 5 + ptile['height'] * 100;

      if (city_worked_positions[ptile['index']] == null && Math.abs(d[0]) <= 2 && Math.abs(d[1]) <= 2) {
        let mesh = new THREE.Mesh( new THREE.RingGeometry( 1, 15, 30), other_city_material );
        city_worked_positions[ptile['index']] = mesh;
        mesh.position.set(pos['x'] - 12, height + 3, pos['y'] - 9);
        mesh.rotation.x = -1 * Math.PI / 2;
        if (scene != null) {
          scene.add(mesh);
        }
      }


    }
  }

}


/****************************************************************************
  Remove all city worked labels.
****************************************************************************/
function remove_city_worked_tiles() {
  for (var workedid in city_worked_positions) {
    if (city_worked_positions[workedid] != null) {
      scene.remove(city_worked_positions[workedid]);
    }
    delete city_worked_positions[workedid];
  }

  if (active_city != null && city_tile(active_city) != null && city_label_positions[city_tile(active_city)['index']] != null) {
      city_label_positions[city_tile(active_city)['index']].visible = true;
  }
}

/****************************************************************************
Is the given tile a city tile in the active city?
****************************************************************************/
function is_city_tile(ptile, active_city)
{
  var ctile = city_tile(active_city);
  var d = map_distance_vector(ptile, ctile);

  if ((d[0] == 2 && d[1] == 2) || (d[0] == -2 && d[1] == -2) || (d[0] == -2 && d[1] == 2) || (d[0] == 2 && d[1] == -2) ) {
    return false;
  }

  if (d[0] <= 2 && d[1] <= 2 && d[0] >= -2 && d[1] >= -2) {
    return true;
  }
  return false;

}

/****************************************************************************
...
****************************************************************************/
function hide_city_labels() {
  if (city_labels_visible) {
    for (var cid in city_label_positions) {
      var city_label = city_label_positions[cid];
      city_label.visible = false;
    }
    city_labels_visible = false;
    $("#city_labels_hide_button").text("Show city labels");
  } else {
    for (let cid in city_label_positions) {
      var city_label = city_label_positions[cid];
      city_label.visible = true;
    }
    city_labels_visible = true;
    $("#city_labels_hide_button").text("Hide city labels");
  }


}

/**************************************************************************
 Returns the city walls scale of for the given city.
**************************************************************************/
function get_citywalls_scale(pcity)
{
  var style = get_citywalls_models(pcity);
  var scale = 9;
  if (pcity['size'] >=3 && pcity['size'] <=6) {
    scale = 9.5;
  } else if (pcity['size'] > 6 && pcity['size'] <= 9) {
    scale = 10.5;
  } else if (pcity['size'] > 9 && pcity['size'] <= 11) {
    scale = 11.0;
  } else if (pcity['size'] > 11) {
    scale = 12.5;
  }
  if (style == "citywalls_stone") {
    return scale;
  } else if (style == "citywalls_roman") {
    return scale * 0.19;
  }
}

/****************************************************************************
...
****************************************************************************/
function get_citywalls_models(pcity) {
  if (pcity['style'] == 0 || pcity['style'] == 5) {
    return "citywalls_stone";
  } else {
    return "citywalls_roman";
  }
}

/****************************************************************************
 Show or hide city buildings on the map.
 ****************************************************************************/
function update_show_city_buildings() {

  if (show_buildings) {
    for (let key in city_building_positions) {
      if (city_building_positions[key] != null) {
        let building = city_building_positions[key];
        building.visible = true;
      }
    }
  } else {
    for (let key in city_building_positions) {
      if (city_building_positions[key] != null) {
        let building = city_building_positions[key];
        building.visible = false;
      }
    }
  }
}

File: freecivx-server/src/main/java/net/freecivx/game/City.java
/**********************************************************************
 Freecivx - the 3D web version of Freeciv. http://www.FreecivWorld.net/
 Copyright (C) 2009-2025  The Freeciv-web project, Andreas Røsdal

 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU Affero General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU Affero General Public License for more details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.

 ***********************************************************************/

package net.freecivx.game;

import java.util.ArrayList;
import java.util.List;

public class City {

    private String name;
    private long owner;
    private long tile;
    private int size;
    private int style;
    private boolean capital;
    private boolean occupied;
    private int walls;
    private boolean happy;
    private boolean unhappy;
    /** IDs of improvements (buildings) present in this city. */
    private List<Integer> improvements = new ArrayList<>();
    private int productionKind;
    private int productionValue;
    /**
     * List of tile indices currently being worked by this city's citizens.
     * The first entry is always the city centre tile; additional entries are
     * assigned as the city grows.  Mirrors the worked-tile list maintained by
     * the C Freeciv server's city struct.
     */
    private List<Long> workedTiles = new ArrayList<>();
    /**
     * Production worklist: a queue of future production targets for this city.
     * Each element is a two-element {@code int[]} where {@code [0]} is the
     * internal production kind (0=unit, 1=improvement) and {@code [1]} is the
     * production target ID.  When the current production is completed the first
     * item in this list is promoted to become the active production target.
     * Mirrors the {@code worklist} field in the C Freeciv city struct.
     */
    private List<int[]> worklist = new ArrayList<>();

    /**
     * Accumulated production shields towards current build target.
     * Mirrors {@code shield_stock} in the C Freeciv city struct.
     */
    private int shieldStock = 0;
    /**
     * Accumulated food in the granary.
     * Mirrors {@code food_stock} in the C Freeciv city struct.
     */
    private int foodStock = 0;

    // Constructor (backwards-compatible: accepts the old String improvements arg and ignores it)
    public City(String name, long owner, long tile, int size, int style, boolean capital, boolean occupied, int walls,
                boolean happy, boolean unhappy, String improvements, int productionKind, int productionValue) {
        this.name = name;
        this.owner = owner;
        this.tile = tile;
        this.size = size;
        this.style = style;
        this.capital = capital;
        this.occupied = occupied;
        this.walls = walls;
        this.happy = happy;
        this.unhappy = unhappy;
        // Old callers passed an empty String; migrate to list representation
        this.productionKind = productionKind;
        this.productionValue = productionValue;
    }

    // Getters and Setters
    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public long getOwner() {
        return owner;
    }

    public void setOwner(long owner) {
        this.owner = owner;
    }

    public long getTile() {
        return tile;
    }

    public void setTile(long tile) {
        this.tile = tile;
    }

    public int getSize() {
        return size;
    }

    public void setSize(int size) {
        this.size = size;
    }

    public int getStyle() {
        return style;
    }

    public void setStyle(int style) {
        this.style = style;
    }

    public boolean isCapital() {
        return capital;
    }

    public void setCapital(boolean capital) {
        this.capital = capital;
    }

    public boolean isOccupied() {
        return occupied;
    }

    public void setOccupied(boolean occupied) {
        this.occupied = occupied;
    }

    public int getWalls() {
        return walls;
    }

    public void setWalls(int walls) {
        this.walls = walls;
    }

    public boolean isHappy() {
        return happy;
    }

    public void setHappy(boolean happy) {
        this.happy = happy;
    }

    public boolean isUnhappy() {
        return unhappy;
    }

    public void setUnhappy(boolean unhappy) {
        this.unhappy = unhappy;
    }

    /** Returns the list of improvement IDs built in this city. */
    public List<Integer> getImprovements() {
        return improvements;
    }

    public void setImprovements(List<Integer> improvements) {
        this.improvements = improvements;
    }

    /** Returns {@code true} if the improvement with the given ID is built here. */
    public boolean hasImprovement(int improvementId) {
        return improvements.contains(improvementId);
    }

    /** Adds an improvement to the city if not already present. */
    public void addImprovement(int improvementId) {
        if (!improvements.contains(improvementId)) {
            improvements.add(improvementId);
        }
    }

    /** Returns the list of tile indices currently worked by this city. */
    public List<Long> getWorkedTiles() {
        return workedTiles;
    }

    /**
     * Registers a tile as being worked by this city.
     * Does nothing if the tile is already in the list.
     *
     * @param tileId the index of the tile to mark as worked
     */
    public void addWorkedTile(long tileId) {
        if (!workedTiles.contains(tileId)) {
            workedTiles.add(tileId);
        }
    }

    /**
     * Removes a tile from this city's worked list.
     *
     * @param tileId the index of the tile to unmark
     */
    public void removeWorkedTile(long tileId) {
        workedTiles.remove(Long.valueOf(tileId));
    }

    /**
     * Returns the production worklist for this city.
     * Each element is {@code int[]{internalKind, productionValue}} where
     * {@code internalKind} is 0 for a unit type and 1 for an improvement.
     *
     * @return the mutable worklist; never {@code null}
     */
    public List<int[]> getWorklist() {
        return worklist;
    }

    /**
     * Replaces the production worklist with the given list.
     *
     * @param worklist the new worklist; must not be {@code null}
     * @throws NullPointerException if {@code worklist} is {@code null}
     */
    public void setWorklist(List<int[]> worklist) {
        this.worklist = java.util.Objects.requireNonNull(worklist, "worklist");
    }

    public int getProductionKind() {
        return productionKind;
    }

    public void setProductionKind(int productionKind) {
        this.productionKind = productionKind;
    }

    public int getProductionValue() {
        return productionValue;
    }

    public void setProductionValue(int productionValue) {
        this.productionValue = productionValue;
    }

    /** Returns the accumulated shields in the production queue. */
    public int getShieldStock() {
        return shieldStock;
    }

    public void setShieldStock(int shieldStock) {
        this.shieldStock = shieldStock;
    }

    /** Returns the food stored in the city granary. */
    public int getFoodStock() {
        return foodStock;
    }

    public void setFoodStock(int foodStock) {
        this.foodStock = foodStock;
    }

    @Override
    public String toString() {
        return "City{" +
                "name='" + name + '\'' +
                ", owner=" + owner +
                ", tile=" + tile +
                ", size=" + size +
                ", style=" + style +
                ", capital=" + capital +
                ", occupied=" + occupied +
                ", walls=" + walls +
                ", happy=" + happy +
                ", unhappy=" + unhappy +
                ", improvements=" + improvements +
                ", productionKind=" + productionKind +
                ", productionValue=" + productionValue +
                ", shieldStock=" + shieldStock +
                '}';
    }
}

